ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -238,6 +292,190 @@ VALUE cpu_and_wall_time_collector_sample(VALUE self_instance) {
   return Qnil;
 }
 
+// This function gets called when Ruby is about to start running the Garbage Collector on the current thread.
+// It updates the per_thread_context of the current thread to include the current cpu/wall times, to be used to later
+// create a stack sample that blames the cpu/wall time spent from now until the end of the garbage collector work.
+//
+// Safety: This function gets called while Ruby is doing garbage collection. While Ruby is doing garbage collection,
+// *NO ALLOCATION* is allowed. This function, and any it calls must never trigger memory or object allocation.
+// This includes exceptions and use of ruby_xcalloc (because xcalloc can trigger GC)!
+//
+// Assumption 1: This function is called in a thread that is holding the Global VM Lock. Caller is responsible for enforcing this.
+void cpu_and_wall_time_collector_on_gc_start(VALUE self_instance) {
+  struct cpu_and_wall_time_collector_state *state;
+  if (!rb_typeddata_is_kind_of(self_instance, &cpu_and_wall_time_collector_typed_data)) return;
+  // This should never fail the the above check passes
+  TypedData_Get_Struct(self_instance, struct cpu_and_wall_time_collector_state, &cpu_and_wall_time_collector_typed_data, state);
+
+  struct per_thread_context *thread_context = get_context_for(rb_thread_current(), state);
+
+  // If there was no previously-existing context for this thread, we won't allocate one (see safety). For now we just drop
+  // the GC sample, under the assumption that "a thread that is so new that we never sampled it even once before it triggers
+  // GC" is a rare enough case that we can just ignore it.
+  // We can always improve this later if we find that this happens often (and we have the counter to help us figure that out)!
+  if (thread_context == NULL) {
+    state->stats.gc_samples_missed_due_to_missing_context++;
+    return;
+  }
+
+  // If these fields are set, there's an existing GC sample that still needs to go out. As a simplification, we just drop
+  // the new GC sample, under the assumption that this won't happen often.
```

Comment:
Did you see this happening, or this more like "we can't fully guarantee this won't happen, but it would be weird if it did".

Reply:
As a follow up, does it make sense for us to reset the values of `cpu_time_at_finish_ns` and `wall_time_at_finish_ns` if we find ourselves in an invalid state?

Reply:
> Did you see this happening, or this more like "we can't fully guarantee this won't happen, but it would be weird if it did".

It's more of the latter, yes. In my older attempt at this (#2304) this couldn't happen, since the plan was to flush the sample immediately when `on_gc_finish` was called.

Because that didn't work out (as explained in the PR description), there is no way to guarantee that Ruby will not trigger more than one GC cycle before we can actually run the `sample_after_gc` method (which we'll queue for execution using `rb_postponed_job_register_one`, which is the same approach we use for regular samples).

But yeah I don't expect that to happen in a normal situation.

> As a follow up, does it make sense for us to reset the values of cpu_time_at_finish_ns and wall_time_at_finish_ns if we find ourselves in an invalid state?

I... don't think so, because I wouldn't consider this case an invalid state -- just a really rare one.

E.g., if we do find out that this is not as rare as we thought, I think the solution is to accumulate multiple samples --- e.g. maybe have extra fields that can collect the total cpu time / wall time that was not flushed yet across multiple GC executions ---  rather than just ignoring the sample that we did not process yet.

Reply:
Makes sense thank you!

Could you capture this in a comment:
> there is no way to guarantee that Ruby will not trigger more than one GC cycle before we can actually run the sample_after_gc method

I didn't think of this case of multiple GCs before `sample_after_gc`, but it's something important to keep in mind in the future.

Reply:
Added in 9477bba5fcdd5b5812ff7ea626814a4e01eaa4a3 (also added a bit more docs)

Reply:
Actually... I found a bit of a bug around this, but decided that it was a feature instead and revised my approach yet again in 872f5af3550a2215631cfdba29f6b91fbbaf42d2.

The bug was when the following call order happened: `on_gc_start` (time=0) -> `on_gc_finish` (time=1) -> `on_gc_start` (time=2) -> `on_gc_finish` (time=3) -> `sample_after_gc`

As we've discussed here, the second `on_gc_start` would be a no-op (since there was a sample to be flushed), but the second `on_gc_finish` actually was not a no-op, because it forgot to test for the same situation.

But because the second `on_gc_finish` clobbered the existing finish timestamps with later ones, the final effect is that the two GCs were "collapsed" into one bigger one in our accounting, e.g. the `sample_after_gc` would use time=0 for start and time=3 for finish.

Which... is actually not bad semantics. If GCs are so back-to-back that we can't even get to `sample_after_gc`, it's kinda irrelevant if in between two GCs we actually still executed a handful Ruby bytecodes or not.

Thus, I've added specs to clarify that this behavior is intended, and updated the comments.